lib/trie/hash_test.go

```diff
@@ -4,72 +4,1009 @@
+	testCases := map[string]struct {
+		n                node
+		writes           []writeCall
+		leafEncodingCopy bool
+		leafBufferLen    int
+		leafBufferBytes  []byte
+		parallel         bool
+		wrappedErr       error
+		errMessage       string
+	}{
+		"branch error": {
+			n: &branch{
+				encoding: []byte{1, 2, 3},
+			},
+			writes: []writeCall{
+				{written: []byte{1, 2, 3}, err: errTest},
+			},
+			wrappedErr: errTest,
+			errMessage: "cannot encode branch: " +
+				"cannot write stored encoding to buffer: " +
+				"test error",
+		},
+		"branch success": {
+			n: &branch{
+				encoding: []byte{1, 2, 3},
+			},
+			writes: []writeCall{
+				{written: []byte{1, 2, 3}},
+			},
+		},
+		"leaf error": {
+			n: &leaf{
+				encoding: []byte{1, 2, 3},
+			},
+			writes: []writeCall{
+				{written: []byte{1, 2, 3}, err: errTest},
+			},
+			wrappedErr: errTest,
+			errMessage: "cannot encode leaf: " +
+				"cannot write stored encoding to buffer: " +
+				"test error",
+		},
+		"leaf success": {
+			n: &leaf{
+				encoding: []byte{1, 2, 3},
+			},
+			writes: []writeCall{
+				{written: []byte{1, 2, 3}},
+			},
+			leafEncodingCopy: true,
+			leafBufferLen:    3,
+			leafBufferBytes:  []byte{1, 2, 3},
+		},
+		"nil node error": {
+			writes: []writeCall{
+				{written: []byte{0}, err: errTest},
+			},
+			wrappedErr: errTest,
+			errMessage: "cannot encode nil node: test error",
+		},
+		"nil node success": {
+			writes: []writeCall{
+				{written: []byte{0}},
+			},
+		},
+		"unsupported node type": {
+			n:          NewMocknode(nil),
+			wrappedErr: ErrNodeTypeUnsupported,
+			errMessage: "node type is not supported: *trie.Mocknode",
+		},
+	}
 
-	if err != nil {
-		t.Errorf("did not hash leaf node: %s", err)
-	} else if buffer.Len() == 0 {
-		t.Errorf("did not hash leaf node: nil")
+	for name, testCase := range testCases {
+		testCase := testCase
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			ctrl := gomock.NewController(t)
+
+			buffer := NewMockbytesBuffer(ctrl)
+			var previousCall *gomock.Call
+			for _, write := range testCase.writes {
+				call := buffer.EXPECT().
+					Write(write.written).
+					Return(write.n, write.err)
+
+				if previousCall != nil {
+					call.After(previousCall)
+				}
+				previousCall = call
+			}
+
+			if testCase.leafEncodingCopy {
+				previousCall = buffer.EXPECT().Len().
+					Return(testCase.leafBufferLen).
+					After(previousCall)
+				buffer.EXPECT().Bytes().
+					Return(testCase.leafBufferBytes).
+					After(previousCall)
+			}
+
+			err := encodeNode(testCase.n, buffer, testCase.parallel)
+
+			if testCase.wrappedErr != nil {
+				assert.ErrorIs(t, err, testCase.wrappedErr)
+				assert.EqualError(t, err, testCase.errMessage)
+			} else {
+				require.NoError(t, err)
+			}
+		})
 	}
 }
 
-func TestHashBranch(t *testing.T) {
-	n := &branch{key: generateRandBytes(380), value: generateRandBytes(380)}
-	n.children[3] = &leaf{key: generateRandBytes(380), value: generateRandBytes(380)}
```

Comment:
should we add a test for random nodes as well?

Reply:
Why? :thinking: 
I'm a big fan of deterministic tests.
Unless we need crypto randomness in certain niche cases (i.e.generate uniform distribution etc.)?